File: ai-service/app/graph.py
from langgraph.graph import StateGraph, END
from app.state import GraphState
from app.nodes import (
    fetch_content_node,
    validate_content_node,
    analyze_content_node,
    score_and_decide_node,
    save_results_node
)

def create_graph():
    workflow = StateGraph(GraphState)

    # Add nodes
    workflow.add_node("fetch", fetch_content_node)
    workflow.add_node("validate", validate_content_node)
    workflow.add_node("analyze", analyze_content_node)
    workflow.add_node("score", score_and_decide_node)
    workflow.add_node("save", save_results_node)

    # Add edges
    workflow.set_entry_point("fetch")
    
    workflow.add_edge("fetch", "validate")
    workflow.add_edge("validate", "analyze")
    workflow.add_edge("analyze", "score")
    workflow.add_edge("score", "save")
    workflow.add_edge("save", END)

    return workflow.compile()

app_graph = create_graph()

File: lib/ai/intelligence/graph.ts
import { StateGraph, END, START } from "@langchain/langgraph";
import { AgentState } from "./state";
import { searchNode } from "./nodes/search";
import { mediaNode } from "./nodes/media";
import { creatorNode } from "./nodes/creator";
import { orchestratorNode } from "./nodes/orchestrator";
import { safetyNode } from "./nodes/safety";
import { analyticsNode } from "./nodes/analytics";
import { learningNode } from "./nodes/learning";

/**
 * Socialyncara Unified Intelligence Pipeline
 *
 * Sequence:
 *   1. research  — Context gathering (topic + audience trend analysis)
 *   2. media     — Media acquisition (conditional on post type)
 *   3. generate  — Human-like content creation
 *   4. orchestrate — Platform-specific tailoring
 *   5. verify    — Safety & compliance check
 *   6. analyze   — Engagement analytics (conditional on safety)
 *   7. learn     — Feedback loop & memory update
 */
const workflow = new StateGraph(AgentState)
    .addNode("research", searchNode)
    .addNode("media", mediaNode)
    .addNode("generate", creatorNode)
    .addNode("orchestrate", orchestratorNode)
    .addNode("verify", safetyNode)
    .addNode("analyze", analyticsNode)
    .addNode("learn", learningNode);

// ── Sequential edges ──────────────────────────────────────────────────────
workflow.addEdge(START, "research");
workflow.addEdge("research", "media");
workflow.addEdge("media", "generate");
workflow.addEdge("generate", "orchestrate");
workflow.addEdge("orchestrate", "verify");

// ── Conditional edge: only proceed to analytics if content is safe ─────────
workflow.addConditionalEdges(
    "verify",
    (state) => state.nextAction === "analyze" ? "analyze" : END
);

workflow.addEdge("analyze", "learn");
workflow.addEdge("learn", END);

export const intelligenceGraph = workflow.compile();
